server/permalinks.go

```diff
@@ -0,0 +1,151 @@
+
+// maxPreviewLines sets the maximum number of preview lines that will be shown
+// while replacing a permalink.
+const maxPreviewLines = 10
+
+// permalinkLineContext shows the number of lines before and after to show
+// if the link points to a single line.
+const permalinkLineContext = 3
+
+// replacement holds necessary info to replace GitLab permalinks
+// in messages with a code preview block.
+type replacement struct {
+	index         int      // index of the permalink in the string
+	word          string   // the permalink
+	permalinkInfo struct { // holds the necessary metadata of a permalink
+		haswww string
+		commit string
+		user   string
+		repo   string
+		path   string
+		line   string
+	}
+}
+
+// getReplacements returns the permalink replacements that need to be performed
+// on a message. The returned slice is sorted by the index in ascending order.
+func (p *Plugin) getReplacements(msg string) []replacement {
+	// find the permalinks from the msg using a regex
+	matches := p.gitlabPermalinkRegex.FindAllStringSubmatch(msg, -1)
+	indices := p.gitlabPermalinkRegex.FindAllStringIndex(msg, -1)
+	var replacements []replacement
+	for i, m := range matches {
+		// have a limit on the number of replacements to do
+		if i > maxPermalinkReplacements {
+			break
+		}
+		word := m[0]
+		index := indices[i][0]
+		r := replacement{
+			index: index,
+			word:  word,
+		}
+		// ignore if the word is inside a link
+		if isInsideLink(msg, index) {
+			continue
+		}
+		// populate the permalinkInfo with the extracted groups of the regex
+		for j, name := range p.gitlabPermalinkRegex.SubexpNames() {
+			if j == 0 {
+				continue
+			}
+			switch name {
+			case "haswww":
+				r.permalinkInfo.haswww = m[j]
+			case "user":
+				r.permalinkInfo.user = m[j]
+			case "repo":
+				r.permalinkInfo.repo = m[j]
+			case "commit":
+				r.permalinkInfo.commit = m[j]
+			case "path":
+				r.permalinkInfo.path = m[j]
+			case "line":
+				r.permalinkInfo.line = m[j]
+			}
+		}
+		replacements = append(replacements, r)
+	}
+	return replacements
+}
+
+// makeReplacements performs the given replacements on the msg and returns
+// the new msg. The replacements slice needs to be sorted by the index in ascending order.
+func (p *Plugin) makeReplacements(msg string, replacements []replacement, glClient *gitlab.Client) string {
+	// iterating the slice in reverse to preserve the replacement indices.
+	for i := len(replacements) - 1; i >= 0; i-- {
+		r := replacements[i]
+		// quick bailout if the commit hash is not proper.
+		if _, err := hex.DecodeString(r.permalinkInfo.commit); err != nil {
+			p.API.LogError("bad git commit hash in permalink", "error", err.Error(), "hash", r.permalinkInfo.commit)
+			continue
+		}
+
+		// get the file contents
+		opts := gitlab.GetFileOptions{
+			Ref: &r.permalinkInfo.commit,
+		}
+		projectPath := fmt.Sprintf("%s/%s", r.permalinkInfo.user, r.permalinkInfo.repo)
+		// TODO: make all of these requests concurrently.
```

Comment:
0/5 Can we please do this now? Especially in the context of my `MessageWillBePosted` remark.

Reply:
Done